File: src/lib/ipc/bennu/sheet.ts
/**
 * A spreadsheet in the project, as a grid.
 *
 * The reading happens in the backend (`bennu-sheet`), not here: a spreadsheet is a container
 * format, and the alternative was a JavaScript parser running in the WebView. What crosses the
 * seam is already **rendered** — the serial number that is a date has become a date — because that
 * conversion is the one piece of real logic in reading a spreadsheet and it belongs where it is
 * tested rather than in a component.
 */

import { bennu } from '../rpc';

/** What a cell **is** — what the grid aligns and tints by. */
export type SheetCellKind = 'empty' | 'text' | 'number' | 'date' | 'bool' | 'error';

/** One cell, already rendered. */
export interface SheetCell {
  text: string;
  kind: SheetCellKind;
}

/** One sheet. Every row is padded to `columns`, so the grid draws a rectangle without measuring. */
export interface Sheet {
  name: string;
  columns: number;
  rows: SheetCell[][];
  /** The sheet has more rows or columns than were read. Said out loud rather than hidden: what is
   *  on screen must never be mistaken for all of it. */
  truncated: boolean;
}

/** A workbook, as much of it as is worth looking at. */
export interface Workbook {
  /** `xlsx` — what the bytes turned out to be, which is not always what the name said. */
  format: string;
  sheets: Sheet[];
  /** The workbook has more sheets than were read. */
  truncated: boolean;
}

/** Read a spreadsheet. Rejects with a sentence worth showing — an unreadable file, a format that
 *  is not supported yet (the old binary `.xls`), or one too large for a viewer.
 *  Wire: `bennu_read_sheet`. */
export function readSheet(file: string): Promise<Workbook> {
  return bennu('bennu_read_sheet', { args: { file } });
}

File: crates/products/bennu/be/src/sheet.rs
//! `bennu_read_sheet` — a spreadsheet in the project, as a grid.
//!
//! Read in the backend rather than in the viewer, and that is the whole design decision here: a
//! spreadsheet is a container format, and the alternative was a JavaScript library that parses it
//! in the WebView. Doing it here means the reading is [tested](bennu_sheet), the bytes cross the
//! seam once as rendered text rather than as a file to re-parse, and the one piece of real logic —
//! a serial number that is actually a date — lives where it can be checked.

use bennu_core::prelude::BennuState;
use bennu_sheet::prelude::Workbook;
use serde::Deserialize;

/// Args for [`bennu_read_sheet`].
#[derive(Deserialize)]
pub struct SheetArgs {
    /// Absolute path of the spreadsheet.
    pub file: String,
}

/// The most bytes read from one spreadsheet.
///
/// A viewer is for looking at a file, and past this the answer to "what is in it" is a tool that
/// filters rather than one that draws. Refused with a sentence rather than by running out of
/// memory quietly.
const MAX_BYTES: u64 = 64 * 1024 * 1024;

/// Read a spreadsheet into sheets of rendered cells.
///
/// The format is decided by the file's own bytes, not by its extension: an `.xls` that is really an
/// OOXML file is something every export tool has shipped, and a reader that trusted the name would
/// refuse a file that opens perfectly. What it turned out to be comes back in `format`.
#[arbor_rpc::handler]
fn bennu_read_sheet(_ctx: &BennuState, args: SheetArgs) -> Result<Workbook, String> {
    let path = std::path::Path::new(&args.file);
    let size = std::fs::metadata(path).map(|m| m.len()).unwrap_or(0);
    if size > MAX_BYTES {
        return Err(format!(
            "that spreadsheet is {} MB — too large to open in a viewer",
            size / (1024 * 1024)
        ));
    }
    let bytes = std::fs::read(path).map_err(|e| format!("could not read {}: {e}", args.file))?;
    bennu_sheet::prelude::read(&bytes)
}
